### ISSUE TITLE
todense() on jax sparse COO returning only zeros

### ISSUE THREAD

**Issue by T-Almeida** [google/jax#18124]
### Description

I was testing the jax.experimental.sparse library and found that todense() method on the COO sparse matrix is returning a vector of zeros.

Code to reproduce (https://colab.research.google.com/drive/1_xxQEqEl5c9NMvYYe8zKsqOS6LJ_V_9P?usp=sharing):

```python
coo_shape = (6,1)
python_coo = {1:10, 3:20, 5:20}

# to array
py_indices = list(python_coo.keys())
py_data = list(python_coo.values())

indices = jnp.array([py_indices, [1]*len(py_indices)]).T
data = jnp.array(py_data, dtype=jnp.float32)

vec_coo = sparse.BCOO((data, indices), shape=coo_shape)

vec_coo.todense() # outputs a vector full of zeros

Array([[0.],
       [0.],
       [0.],
       [0.],
       [0.],
       [0.]], dtype=float32)
```

Currently, I am using a workaround with jax.lax.scatter:

```python
def vec_coo_todense_wscatter(coo):
  indices = coo.indices[:,0]

  dnums = jax.lax.ScatterDimensionNumbers(update_window_dims=(), inserted_window_dims=(0,), scatter_dims_to_operand_dims=(0,))
  return jax.lax.scatter(jnp.zeros((coo.shape[0],)), jnp.expand_dims(indices, axis=-1), coo.data, dimension_numbers=dnums)

# returns a dense vector with the correct data
vec_coo_todense_wscatter(vec_coo)

Array([ 0., 10.,  0., 20.,  0., 20.], dtype=float32)

```

### What jax/jaxlib version are you using?

v0.4.16

### Which accelerator(s) are you using?

Tested on CPU and GPU

**Comment by T-Almeida** [comment]
Update:

The "bug" that I was experience was related to my indices and the shape of the matrix i believe:

```python
indices = jnp.array([py_indices, [1]*len(py_indices)]).T
shape = (6,1)
```
As you can see I am wrongly setting the column indices to "1" instead of "0". 

Should this be impossible? I am trying the index column 1, but according to the shape thats impossible. This should thrown an error?

**Comment by jakevdp** [comment]
I think you meant to use this:
```python
indices = jnp.array([py_indices, [0]*len(py_indices)]).T
```
For an axis of size 1, an index value 1 is out-of-bounds (indexing in Python is generally zero-based), and BCOO uses out-of-bound indexing to indicate padded values that don't affect the array.

**Comment by T-Almeida** [comment]
Exactly, @jakevdp. As mentioned above, I realized that later. However, shouldn't jax have thrown an error by validating out-of-bounds?

**Comment by jakevdp** [comment]
As I mentioned, JAX's `BCOO` matrix uses out-of-bound indices to mark padded values in its representation, so the data you passed is valid, it's just that it represents a different matrix than you intended it to represent.

**Comment by T-Almeida** [comment]
Ty for the explanation :) Really appreciated the link. 

Closing the issue
